### ISSUE TITLE
some Go tests fail due to missing local temp directories

### ISSUE THREAD

**Issue by chrisd8088** [git-lfs/git-lfs#4046]
**Describe the bug**

When running the Go test suite on some development systems (macOS, at least) with `make test`, some tests fail with the errors of the form:
```
    testutils.go:228: Error running git command 'git push origin ...': exit status 1 EOF
        Uploading LFS objects: ...
        error: failed to push some refs to '/var/folders/.../T/lfsRepo...'
```

Running `make test` again with the `GIT_TRACE=1` environment variable yields the information that the given path plus `/.git` does not exist as a local directory.  As an example:
```
12:21:40.442232 trace git-lfs: xfer[git-lfs]: error creating handler: chdir /var/folders/f4/3tj6q1w11xd4mnnx7fm3jk8m0000gn/T/lfsRepo758791364/.git: no such file or directory
        EOF
        Uploading LFS objects: 100% (2/2), 45 B | 0 B/s, done.
        error: failed to push some refs to '/var/folders/f4/3tj6q1w11xd4mnnx7fm3jk8m0000gn/T/lfsRepo758791364'
```

**To Reproduce**
Steps to reproduce the behaviour:
1. Build from source per the Wiki [instructions](https://github.com/git-lfs/git-lfs/wiki/Installation#source).
2. Run the Go test suite with `make test` (not the shell tests in the `t/` directory).
3. Observe the `TestCurrentRefAndCurrentRemoteRef`, `TestRecentBranches`, and `TestScanUnpushed` tests fail with errors as described above.

**Expected behaviour**
All Go tests should pass.

**System environment**
macOS Catalina with Go version 1.13.3.

**Output of `git lfs env`**
Not applicable.

**Additional context**
Not applicable.

**Comment by bk2204** [comment]
Can you try the `standalone-file` branch on my remote and see if this fixes it for you?  It looks like this passes on macOS on the test servers, but it isn't clear to me why it's not passing for you on your system.

**Comment by chrisd8088** [comment]
Hmm.  Unfortunately, on a quick trial, I get the same results on the Mac:
```
22:02:21.196964 trace git-lfs: xfer[git-lfs]: error creating handler: chdir /var/folders/f4/3tj6q1w11xd4mnnx7fm3jk8m0000gn/T/lfsRepo915408172/.git: no such file or directory
EOF
Uploading LFS objects: 100% (3/3), 71 B | 0 B/s, done.
error: failed to push some refs to '/var/folders/f4/3tj6q1w11xd4mnnx7fm3jk8m0000gn/T/lfsRepo915408172'
```
I can try to look into this a bit more over the next couple days too.

**Comment by chrisd8088** [comment]
So I spent some time on this, slowing digging through various misunderstandings on my part, and I think there are two things going on for me on the Mac.

The first is that because the test suite seems to [execute](https://github.com/git-lfs/git-lfs/blob/e8be7992b1c42956a351abf876e7c137b2254d83/tq/custom.go#L131-L132) the standalone transfer adapter as a separate process by invoking [`git-lfs standalone-file`](https://github.com/git-lfs/git-lfs/blob/e8be7992b1c42956a351abf876e7c137b2254d83/tq/custom.go#L358).  That by default invokes the globally installed `git-lfs`, not the one that's been compiled locally in the working tree.

So after defeating that with a simple `PATH` override, the second issue is that I now get (from your `standalone-failure` branch) a different failure:
```
12:15:33.605134 trace git-lfs: tq: starting transfer adapter "lfs-standalone-file"
12:15:33.636535 trace git-lfs: xfer[git-lfs]: error creating handler: failed to call git rev-parse --git-dir: exit status 128
```

After stubbing in some crude debugging code, it looks to me as though the directory [passed](https://github.com/git-lfs/git-lfs/blob/e8be7992b1c42956a351abf876e7c137b2254d83/lfshttp/standalone/standalone.go#L92) in the `path` argument to `gitDirAtPath()` doesn't exist, namely, the `/var/folders/.../lfsRepo.../.git` one.  But, the parent directory (`lfsRepo...`) does exist.  So perhaps the immediate fix would be to strip off any `/.git` that's trailing at the end of the `path` argument.

I would also love to find a way to have the Go test suite ensure that it's only executing the locally-compiled binaries, and not whatever happens to be found first in the default `PATH`.

If you have any comments on either item, let me know, and in the meantime I'll take a stab at trying to implement one or both.

**Comment by bk2204** [comment]
It's curious to me that we're adding `.git` on the end, because we should be getting a valid remote from the other side.  If you can figure out why that's happening, I'd be very interested.

I think for the `PATH` issue, it might be sufficient to just include our `bin` directory at the front, which should make things work out of the box.

**Comment by chrisd8088** [comment]
Yeah ... more study is needed, too, because while the latter problem was reproducible for a while this morning, now it appears gone.  That makes me suspicious of some kind of race in setting up the temp repos and starting up the standalone transfer agent, but I will have to investigate further.  Most odd.

**Comment by chrisd8088** [comment]
Just some interim notes -- first, I'm still looking into why the `.git` directory doesn't exist at the point where `gitDirAtPath()` is called in these tests; from what I can see, it should be [created](https://github.com/git-lfs/git-lfs/blob/e8be7992b1c42956a351abf876e7c137b2254d83/lfs/scanner_git_test.go#L20) at the start of the test.

Second, the reason the `/.git` is appended to the remote's "URL", as [returned](https://github.com/git-lfs/git-lfs/blob/e8be7992b1c42956a351abf876e7c137b2254d83/lfsapi/endpoint_finder.go#L120-L122) by `RemoteEndpoint()` [based](https://github.com/git-lfs/git-lfs/blob/e8be7992b1c42956a351abf876e7c137b2254d83/lfsapi/endpoint_finder.go#L135-L137) on the `remote.origin.url` config value (which for these tests is just a path like `/var/folders/.../lfsRepo...`) is that it gets modified by the `NewEndpoint()` [call](https://github.com/git-lfs/git-lfs/blob/e8be7992b1c42956a351abf876e7c137b2254d83/lfsapi/endpoint_finder.go#L148) in `NewEndpointFromCloneURL()`, which in turn [calls](https://github.com/git-lfs/git-lfs/blob/e8be7992b1c42956a351abf876e7c137b2254d83/lfsapi/endpoint_finder.go#L173-L175) `EndpointFromLocalPath()`, which [appends](https://github.com/git-lfs/git-lfs/blob/e8be7992b1c42956a351abf876e7c137b2254d83/lfshttp/endpoint.go#L111-L113) the `/.git`.

**Comment by chrisd8088** [comment]
Update to last night's not -- the `.git/` directory doesn't exist because these are the test's [remote](https://github.com/git-lfs/git-lfs/blob/e8be7992b1c42956a351abf876e7c137b2254d83/lfs/scanner_git_test.go#L55) repos, not the principal test repo, and as such they are [created](https://github.com/git-lfs/git-lfs/blob/e8be7992b1c42956a351abf876e7c137b2254d83/t/cmd/util/testutils.go#L434-L436) as bare repositories.  So that's the fundamental mismatch; there is no `.git/` subdirectory for these, so what `EndpointFromLocalPath()` does in appending that path segment doesn't work for them.

I also realize what I thought was transient behaviour was actually due to the fact that `make test` is actually [running](https://github.com/git-lfs/git-lfs/blob/e8be7992b1c42956a351abf876e7c137b2254d83/Makefile#L409) the Go tests in [package mode](https://golang.org/cmd/go/#hdr-Test_packages), which caches the results.  Sigh.  So forcing it to skip caching with `-count=1` solves that mystery.

With that all said, I think I now have a handle on what's going on here:
1) The `make test` target runs the Go tests in package mode, with caching of results by default.
2) The `make test` target doesn't adjust `$PATH`, so that the `git-lfs` binary that's found may not be the locally compiled one.
3) Certain tests (namely the three identified in this issue) use the `AddRemote()` function, which creates bare repos, and then this conflicts with the endpoint finder logic which adds a `/.git` path segment to local paths.  That leads to problems in the `gitDirAtPath()` function in the standalone transfer adapter.

**Comment by bk2204** [comment]
This is a great analysis.  Since some people may invoke `go test ./...` instead of `make test`, we'll likely need to update the relevant Go tests themselves with the proper `PATH` setting.

And as for `EndpointFromLocalPath`, I think we can just drop the appending of `.git`, since we'll figure it out correctly either way.  That code was implemented in acac3c458 before we had the standalone transfer agent, so it was necessary to make things work with the standalone agents people were using, but I don't think it's needed anymore with the builtin behavior.

**Comment by chrisd8088** [comment]
OK, thanks for the input ... I was heading toward futzing with `$PATH` in the `Makefile`, but let me look into doing it in the relevant Go tests instead.  And I'll try just pulling out the appending of `/.git` too.
